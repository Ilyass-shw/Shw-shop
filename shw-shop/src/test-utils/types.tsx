import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';

export type StoreType = ReturnType<typeof configureStore>;
export type RenderReturn = ReturnType<typeof render>;
export type SliceParameters = Parameters<typeof createSlice>;
export type UiType = React.ReactElement;

export interface WrapperParams {
  MyStore: StoreType;
  WithRouter?: boolean;
}

export interface useFormProps {
  email: string;
  phone: number;
  buyerAcceptMarketing: string;
}
