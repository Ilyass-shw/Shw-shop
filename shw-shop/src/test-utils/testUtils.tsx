import React, { ReactElement } from 'react';
import { render as rtlRender, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import store, { makeCustomTestStore } from './faksestore';
import { RenderReturn, StoreType, useFormProps } from './types';
import { BrowserRouter } from 'react-router-dom';
import { realStoreType } from '../App/store';
import { useForm, FormProvider } from 'react-hook-form';

export const render = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(ui);
};

// ============= ============= ============= ============= =============

export const renderWithStore = (
  ui: React.ReactElement,
  MyStore: realStoreType,
  WithRouter?: 'withRouter',
): RenderReturn => {
  if (WithRouter) {
    return rtlRender(
      <Provider store={MyStore}>
        <BrowserRouter>{ui}</BrowserRouter>
      </Provider>,
    );
  } else {
    return rtlRender(<Provider store={MyStore}>{ui}</Provider>);
  }
};

// ============= ============= ============= ============= =============

export const renderWithRouter = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(<BrowserRouter>{ui}</BrowserRouter>);
};

// ============= ============= ============= ============= =============

export const makeTestStore = (
  Name?: string,
  initialState?: Record<string, unknown>,
  reducers = {},
): StoreType => {
  if (Name && initialState) {
    const CustomTestStore = makeCustomTestStore(Name, initialState, reducers);
    const origDispatch = CustomTestStore.dispatch;
    CustomTestStore.dispatch = jest.fn(origDispatch);
    return CustomTestStore;
  } else {
    const origDispatch = store.dispatch as jest.Mock;
    store.dispatch = jest.fn(origDispatch);
    return store;
  }
};

// re-export everything
export * from '@testing-library/react';

// ============= ============= ============= ============= =============

export const renderWithFormProvider = (
  ui: ReactElement,
  WithRouter?: 'withRouter',
  { defaultValues = {} } = {},
): RenderResult => {
  // let reactHookFormMethods = {};

  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  };
  const WithRouterWrapper: React.FC = ({ children }) => {
    const methods = useForm({ defaultValues });
    return (
      <BrowserRouter>
        <FormProvider {...methods}>{children}</FormProvider>
      </BrowserRouter>
    );
  };

  if (!WithRouter) {
    return rtlRender(ui, { wrapper: Wrapper });
  }
  return rtlRender(ui, { wrapper: WithRouterWrapper });
};

// ============= ============= ============= ============= =============

export const renderWithReactHookForm = (
  ui: JSX.Element,
  saveData: jest.Mock,
): RenderResult => {
  const Wrapper: React.FC = ({ children }) => {
    const methods = useForm<useFormProps>();
    return (
      <FormProvider {...methods}>
        <form
          onSubmit={methods.handleSubmit((data) => {
            saveData(data);
          })}
        >
          <BrowserRouter>{children}</BrowserRouter>
          <button type="submit" data-testid="testSubmitButton"></button>
        </form>
      </FormProvider>
    );
  };
  return rtlRender(<Wrapper>{ui}</Wrapper>);
};
