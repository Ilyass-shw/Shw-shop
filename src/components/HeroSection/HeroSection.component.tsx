import styled from 'styled-components';

export const BackgroundImage = styled.section`
  background: url('https://jonnajintonsweden.com/wp-content/uploads/2020/09/test-header-4-4500x1550.jpg')
    45% / cover no-repeat;
  height: 22.2rem;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-top: 6rem;

  @media screen and (min-width: 520px) {
    background-position: 25% 25%;
  }
`;

// ============= ============= ============= ============= =============

export const HeroContent = styled.div`
  width: 40vw;
  height: 60%;
  color: #f1f1f1;
`;

// ============= ============= ============= ============= =============

export const HeroText = styled.div`
  @import url('https://fonts.googleapis.com/css2?family=Quattrocento+Sans:ital@1&display=swap');

  margin-left: 3vw;
  margin-top: 3vw;

  p {
    padding: 10px 10px 5px 10px;
    font-family: 'Quattrocento Sans', sans-serif;
    font-style: italic;
    text-shadow: 0 0 white;
    font-size: clamp(14px, 17px, 17px);
    width: 55vw;
    max-width: 23.4rem;
  }

  @media screen and (min-width: 364px) and (max-width: 485px) {
    margin-left: 6vw;
    p {
      width: 42vw;
    }
  }
`;

export const FirstPart = styled.p``;

// ============= ============= ============= ============= =============

export const LastPart = styled.p`
  text-align: end;
`;

// ============= ============= ============= ============= =============
