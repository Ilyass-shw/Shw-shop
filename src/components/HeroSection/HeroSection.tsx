import React from 'react';
import {
  BackgroundImage,
  HeroContent,
  HeroText,
  FirstPart,
  LastPart,
} from './HeroSection.component';

const HeroSection: React.FC = () => {
  return (
    <BackgroundImage>
      <HeroContent>
        <HeroText>
          <FirstPart>“Art is like a dance of the Heart,</FirstPart>
          <LastPart> when the Soul is free…”</LastPart>
        </HeroText>
      </HeroContent>
    </BackgroundImage>
  );
};

export default HeroSection;
