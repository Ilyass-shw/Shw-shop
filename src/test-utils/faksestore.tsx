import { configureStore } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import { StoreType } from './types';

const fakeSlice = createSlice({
  name: 'name',
  initialState: {},
  reducers: {},
});

export default configureStore({
  reducer: {
    posts: fakeSlice.reducer,
  },
});

export const makeCustomTestStore = (
  Name: string,
  initialState: Record<string, unknown>,
  reducers = {},
): StoreType => {
  const FakeSlice = createSlice({
    name: Name,
    initialState,
    reducers,
  });

  return configureStore({
    reducer: {
      posts: FakeSlice.reducer,
    },
  });
};
