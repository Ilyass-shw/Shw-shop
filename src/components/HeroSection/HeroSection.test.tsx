import React from 'react';
import { render } from '../../test-utils/testUtils';
import HeroSection from './HeroSection';

describe('HeroSection', () => {
  it('should render', () => {
    render(<HeroSection />);
  });
  it('should match snapshot', () => {
    const { asFragment } = render(<HeroSection />);
    expect(asFragment()).toMatchSnapshot();
  });
});
