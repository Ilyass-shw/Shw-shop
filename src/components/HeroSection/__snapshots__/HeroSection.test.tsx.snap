// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroSection should match snapshot 1`] = `
<DocumentFragment>
  <section
    class="sc-bdvvaa lchavo"
  >
    <div
      class="sc-gsDJrp gfkuTp"
    >
      <div
        class="sc-dkPtyc UlDpb"
      >
        <p
          class="sc-hKwCoD"
        >
          “Art is like a dance of the Heart,
        </p>
        <p
          class="sc-eCImvq iMSVxU"
        >
           when the Soul is free…”
        </p>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
